### PR TITLE
Convert piles from concatenated to stacked-and-padded, and revive existing tests

### PR DESCRIPTION
**Pull request by axch**

### CODE REVIEW

File: jax/_src/core.py [hunk 1]
```diff
@@ -1729,6 +1732,8 @@ def __eq__(self, other):
     if isinstance(other, DArray) and self._aval == other._aval:
       return self._data == other._data
     return False
+  def __len__(self):
```
Comment (mattjj) [inline jax/_src/core.py:1735]: Do we need this? If not, can we remove it?

File: jax/_src/interpreters/batching.py [hunk 2]
```diff
@@ -52,7 +51,8 @@ class PileTy:
   length: Union[int, Tracer, core.Var]
   elt_ty: core.DShapedArray
   def __repr__(self) -> str:
-    return f'Var{id(self.binder)}:{self.length} => {self.elt_ty}'
+    # return f'Var{id(self.binder)}:{self.length} => {self.elt_ty}'
+    return f'i:{self.length} => {self.elt_ty}'
```
Comment (mattjj) [inline jax/_src/interpreters/batching.py:55]: We probably want to revert this change, right?

File: jax/_src/interpreters/batching.py [hunk 3]
```diff
@@ -61,7 +61,7 @@ class IndexedAxisSize:
   idx: core.Var
   lengths: Union[Array, core.Var, Tracer]
   def __repr__(self) -> str:
-    return f'{str(self.lengths)}.Var{id(self.idx)}'
+    return f'{str(self.lengths)}.i'
```
Comment (mattjj) [inline jax/_src/interpreters/batching.py:64]: Also revert this change?

File: jax/_src/interpreters/batching.py [hunk 4]
```diff
@@ -376,12 +380,9 @@ def process_call(self, call_primitive, f, tracers, params):
     sizes = (x.shape[d] if type(d) is int else len(d.segment_lengths)
              for x, d in zip(vals, dims) if d is not not_mapped)
     axis_size, = core.dedup_referents(sizes)
-    segment_lens, dims = unpack_concat_axes(dims)
     f_, dims_out = batch_subtrace(f, self.main, tuple(dims))
-    f_ = _update_annotation(f_, f.in_type, axis_size, self.axis_name, dims,
-                            segment_lens)
-    vals_out = call_primitive.bind(f_, *segment_lens, *vals, **params)
-    vals_out, dims_out = reassemble_concat_axes(vals_out, dims_out())
+    f_ = _update_annotation(f_, f.in_type, axis_size, self.axis_name, dims, [])
```
Comment (mattjj) [inline jax/_src/interpreters/batching.py:384]: This doesn't seem right in general; can you explain why we can always pass an empty list here, and skip the `unpack_concat_axis` / `reassemble_concat_axis` stuff? I assume we can't skip those things in general, but we don't have a test that's exercising this path. If so, we should figure out what needs to be done and leave a TODO (or else fix up this section).

File: jax/_src/interpreters/batching.py [hunk 5]
```diff
@@ -98,23 +98,28 @@ def _pile_unflatten(aval, x):
...
+  del shape[0]
+  shape[ragged_axis-1] = IndexedAxisSize(binder, segment_lens)
   elt_ty = core.DShapedArray(tuple(shape), x.dtype, x.weak_type)
   return Pile(PileTy(binder, axis_size, elt_ty), x)
 
+
 @dataclasses.dataclass(frozen=True)
-class ConcatAxis:
-  axis: int
+class RaggedAxis:
```
Comment (mattjj) [inline jax/_src/interpreters/batching.py:113]: Could you leave a TODO here, or in the PR message, explaining how we need to generalize `RaggedAxis` further, to handle piles like `i:(Fin 3) => f32[lens1.i, lens2.i]` (i.e. multiple ragged axes)?

Comment (axch) [reply jax/_src/interpreters/batching.py:112]: Done.

File: jax/_src/interpreters/batching.py (hunk 4 above)
Comment (axch) [reply jax/_src/interpreters/batching.py:385]: Let's work this one out synchronously.  We don't need to call a function literally called `unpack_concat_axis` because there are no more concat axes; whatever translation this was trying to do should be simpler in the padded-and-stacked representation.  That said, I'm not sure what translation this was trying to do, and what test would exercise it.  I passed an empty list because that's what `unpack_concat_axis` would do on a non-ragged array, which got the existing tests passing.

File: jax/_src/interpreters/batching.py (hunk 3 above)
Comment (axch) [reply jax/_src/interpreters/batching.py]: Done.

File: jax/_src/interpreters/batching.py (hunk 2 above)
Comment (axch) [reply jax/_src/interpreters/batching.py]: Done.

File: jax/_src/core.py (hunk 1 above)
Comment (axch) [reply jax/_src/core.py:1735]: Actually, we seem to -- the `segment_lengths` can be a `DArray`, and we query its `len`.

**Review by mattjj — APPROVED** [review]
So sick, also the PRs to come.

File: jax/_src/interpreters/batching.py (hunk 4 above)
Comment (axch) [reply jax/_src/interpreters/batching.py:385]: In offline discussion, we determined: (i) this code seems to only relevant for getting axes to work under `jit`, which is why current tests pass without it, and (ii) we will revive an updated version in a future PR.
